Annotations

juliaogris opened the pull request:
Fixes #24 .

Changes proposed in this pull request:
- Add syntax `*:` in addition to ` .. * <- * :` for annotations
- Remove pdb invocations
- Add unittests for annotations
- Add e2e (end-to-end) tests for annotations

@anz-bank/sysl-developers

juliaogris commented on src/sysl/core/syslloader.py:
```diff
@@ -309,6 +305,14 @@ def infer_expr_type(expr, top=True):
         infer_expr_type(v.expr)
 
 
+def infer(module):
```
I needed this a public function to create true unit tests without accessing input files.
I'm mot sure if this is the most suited function name - I'm happy to take suggestions.

marcelocantos replied:
`postprocess` should do.

juliaogris replied:
Done

juliaogris commented on src/sysl/util/file.py:
```diff
@@ -22,3 +23,10 @@ def java(self, w, name, package):
         self(w, os.path.join(self.package.replace('.', '/'), name + u'.java'))
+
+
+def areFilesIdentical(fname1, fname2):
```
This function should probably live with a utility module in the tests top-level directory.
However that would require for tests to have a package structure and be on the python path which I wanted to avoid.

marcelocantos replied:
It's fine where it is, but it should be called `filesAreIdentical`. Predicate-style functions should be phrased as logical propositions, not as questions.

juliaogris replied:
Done

juliaogris commented on test/e2e/expected_output/001_annotations.txt:
```diff
@@ -0,0 +1,124 @@
+apps {
```
New e2e tests work by creating `actual_output/XXX.txt` from `input/XXX.syl` and comparing it to `expected_output/XXX.txt`

marcelocantos replied:
Should go into /tmp, but that can be a low-pri issue for a future PR.

juliaogris replied:
I assume you mean the `actual_output` directory should go into `/tmp`?
That wouldn't work on Windows, so I would have to request a Python tmp-dir which is random / different each time.
I've added `tests/e2e/actual_output` to `.gitignore` so it doesn't pollute the repo. The reason why I like `expected_output` and `actual_output` side by sider is that I can use a dir diff tool more easily this way I find.

What's the reason for you suggesting it should go into `/tmp`?

marcelocantos replied:
I prefer to keep the source tree empty of uncommitted output. It's a pretty minor thing, so let's park it for now and discuss later.

juliaogris replied:
ok

juliaogris commented on test/sysl/test_epa.py:
```diff
@@ -148,7 +146,7 @@ def test_ignore_keyword(self):
             self.fail(traceback.format_exc())
 
-    def test_labels(self):
+    def test_labelsX(self):
```
Remove this.

juliaogris commented on test/sysl/test_epa.py:
```diff
@@ -166,6 +164,7 @@ def test_labels(self):
             out = self.integration_view_helper(
                 '/test/data/test_epa_ignore_keyword', d)
 
+            #print out[0]
```
Remove

marcelocantos commented on test/sysl/test_epa.py:
```diff
@@ -287,7 +277,6 @@ def test_missing_labels(self):
     #     out = self.integration_view_helper('/test/data/test_epa_events', d)
 
-    #     #import pdb; pdb.set_trace()
     #     self.assertTrue('state "**Test EPA :: App1 -> App1 Event**" as _0' in out[0])
```
Remove all comments or prefix with a TODO.

juliaogris replied:
Done

marcelocantos commented on src/sysl/util/file.py:
```diff
@@ -22,3 +23,10 @@ def java(self, w, name, package):
+    with open(fname1, "r") as f1:
+        with open(fname2, "r") as f2:
+            return all(line1 == line2
+                       for line1, line2 in izip(f1, f2))
```
Use `izip_longest`.

juliaogris replied:
Nice catch. Done.

marcelocantos commented on test/e2e/test_e2e.py:
```diff
@@ -0,0 +1,36 @@
+	call(cmd)
+
+	expected_fname = path.join(dname, 'expected_output', fname + '.txt')
+	assert areFilesIdentical(expected_fname, out_fname)
```
Replace tabs with spaces.

juliaogris replied:
Oups, I surprised the linter didn't find this.
Done.

juliaogris replied:
This seemed really fishy, so I found that `flake8` doesn't not execute properly when invoked as `python setup.py lint`. It is a bit of a mystery to my why this happens so I've removed it entirely from setup.py and adjusted the CI scripts accordingly.
I've also fixed all new linter warnings that have slipped in in the meantime.

juliaogris commented on src/sysl/core/syslloader.py:
```diff
@@ -309,6 +305,14 @@ def infer_expr_type(expr, top=True):
         infer_expr_type(v.expr)
 
 
+def infer(module):
```
Done

juliaogris commented on src/sysl/exporters/xml/serializer.py:
```diff
@@ -172,8 +172,7 @@ def serializer(context):
                                     if jfname == 'index':
-                                        import pdb
-                                        pdb.set_trace()
+                                        raise Exception('Bad Java name (index)')
```
Done

juliaogris commented:
Something is wrong with the `flake8` runner for the `setup.py` and how it parses `setup.cfg`.
As a consequence `python setup.py lint` didn't do it's job, but `flake8` still works as expected.
In the last commit I've updated the CI scripts to work direcgtly with `pytest` and `flake8` and I've also fixed all the linter warnings that have slipped in previous commits.
